### ISSUE TITLE
[BUG] [EVALS] GenerationMetricsCallback: three eval-time bugs causing degraded eval metrics on chat models

### ISSUE THREAD

**Issue by ronitagarwala01** [RapidFireAI/rapidfireai#269]
## Bug Description
 
`GenerationMetricsCallback` in `rapidfireai/fit/ml/callbacks.py` has three eval-time bugs that cause significant degradation of generated outputs when used with instruction-tuned chat models and prompts longer than 512 tokens. They surface together in my use case but are logically independent:
 
**Bug 1** — `_generate_batch` calls the tokenizer with `max_length=512, truncation=True`. Hardcoded. Any prompt longer than 512 tokens is silently truncated.
 
**Bug 2** — `_prepare_data` calls `self.tokenizer.apply_chat_template(item["prompt"], tokenize=False)` without `add_generation_prompt=True`. For chat templates that gate generation on an explicit assistant-turn opener (e.g. Gemma's `<start_of_turn>model\n`, Qwen's `<|im_start|>assistant\n`), the model receives a prompt ending mid-user-turn instead of at the start of an assistant turn.
 
**Bug 3** — `_generate_batch` calls the tokenizer with `padding=True` (which produces an `attention_mask`) but returns only `input_ids`. The mask is discarded. `_run_generation_loop` then calls `model.generate(input_ids_batch, **gen_config)` with no `attention_mask` argument, so generation attends to pad tokens as if they were context. The padding side is also not explicitly set to `"left"`, which is required for batched generation with prompts of different lengths.
 
## To Reproduce
 
1. SFT an instruction-tuned chat model whose templated prompt exceeds 512 tokens (I reproduced with `Qwen/Qwen2.5-1.5B-Instruct`).
2. Build an `RFSFTConfig` with a user-defined `compute_metrics` that inspects the generated predictions.
3. Run `experiment.run_fit(...)` and inspect the outputs reaching `compute_metrics`.
Generated outputs are truncated, do not begin where an instruction-tuned model would normally begin its turn, and contain off-task text. The same base model + adapter loaded directly via `transformers` and called with `apply_chat_template(..., add_generation_prompt=True)` and an explicit `attention_mask` produces coherent outputs.
 
## Expected Behavior
 
`GenerationMetricsCallback` should produce outputs consistent with running the same model + adapter through `transformers` directly at inference time:
 
- Prompts should not be silently truncated at 512 tokens.
- `apply_chat_template` should be called with `add_generation_prompt=True`.
- The `attention_mask` from the tokenizer should be threaded through to `model.generate`, with left-side padding for batched generation.
## Screenshots
 
N/A.
 
## Environment
 
- Ran on DSMLP container
- Python version: 3.12.13
- RapidFire AI version: 0.15.2
## Additional Context
 
This was flagged to Arun via Piazza; he asked me to file the issue here for tracking. Happy to open a PR when I get more time if maintainers are interested — let me know the preferred design for the `max_length` knob (default to `self.tokenizer.model_max_length`, accept a config-level override, both).
 
## Error Logs
 
Not a traceback — outputs are silently garbled. Below is the combined monkey-patch I've been using as a workaround.
 
<details>
<summary>Click to expand: combined monkey-patch</summary>

```python
from rapidfireai.fit.ml import callbacks as rf_cb
import torch

def _patched_prepare_data(self, eval_dataset):
    input_texts, references = [], []
    for item in eval_dataset:
        if not isinstance(item, dict):
            continue
        if "prompt" in item and "completion" in item:
            reference = item["completion"][-1]["content"]
            input_text = self.tokenizer.apply_chat_template(
                item["prompt"],
                tokenize=False,
                add_generation_prompt=True,   # Bug 2 fix
            )
        elif "input" in item and "output" in item:
            input_text, reference = item["input"], item["output"]
        else:
            continue
        if input_text and reference:
            input_texts.append(input_text)
            references.append(reference.strip())
    return input_texts, references

def _patched_generate_batch(self, model, input_texts):
    if not input_texts:
        self._last_attention_mask = None
        return torch.empty((0, 0), dtype=torch.long).to(model.device)
    self.tokenizer.padding_side = "left"                              # Bug 3 fix (a)
    if self.tokenizer.pad_token is None:
        self.tokenizer.pad_token = self.tokenizer.eos_token
    if getattr(model.config, "pad_token_id", None) is None:
        model.config.pad_token_id = self.tokenizer.pad_token_id
    inputs = self.tokenizer(
        input_texts, return_tensors="pt", padding=True,
        truncation=True, max_length=4096,                             # Bug 1 fix
    ).to(model.device)
    self._last_attention_mask = inputs["attention_mask"]              # Bug 3 fix (b): preserve mask
    return inputs["input_ids"]

def _patched_run_generation_loop(
    self, model, input_ids, batch_references, device,
    override_batch_size=None, generation_config=None, eos_token_ids=None,
):
    predictions, references = [], []
    num_samples = input_ids.shape[0]
    batch_size = override_batch_size or self.batch_size
    gen_config = generation_config if generation_config is not None else self.generation_config
    attn_mask = getattr(self, "_last_attention_mask", None)           # Bug 3 fix (c): use mask

    with torch.no_grad():
        for i in range(0, num_samples, batch_size):
            input_ids_batch = input_ids[i:i+batch_size].to(device)
            kwargs = dict(gen_config)
            if attn_mask is not None:
                kwargs["attention_mask"] = attn_mask[i:i+batch_size].to(device)
            kwargs.setdefault("pad_token_id", self.tokenizer.pad_token_id)

            with torch.inference_mode(), torch.amp.autocast("cuda"):
                outputs_batch = model.generate(input_ids_batch, **kwargs)

            new_tokens = outputs_batch[:, input_ids_batch.shape[1]:]
            if eos_token_ids:
                new_tokens = self._truncate_at_eos(new_tokens, eos_token_ids)
            generated_texts = self.tokenizer.batch_decode(new_tokens, skip_special_tokens=True)
            predictions.extend(generated_texts)
            references.extend(batch_references[i:i+batch_size])
            del outputs_batch, input_ids_batch, new_tokens
            torch.cuda.empty_cache()
    torch.cuda.empty_cache()
    return predictions, references

rf_cb.GenerationMetricsCallback._prepare_data        = _patched_prepare_data
rf_cb.GenerationMetricsCallback._generate_batch      = _patched_generate_batch
rf_cb.GenerationMetricsCallback._run_generation_loop = _patched_run_generation_loop
```
 
</details>
